### PR TITLE
Don't fail to suppress on non-class type definitions (interfaces, enums, records etc)

### PR DESCRIPTION
**Pull request by CRogers**
## Before this PR
Suppressing fails when there is an errorprone error raised on an interface, enum, record etc (anything that isn't a class):

```
public interface InterfaceType {
       ^
     Please report this at https://github.com/google/error-prone/issues/new and include the following:
  
     error-prone version: 2.31.0
     BugPattern: SafeLoggingPropagation
     Stack Trace:
     java.lang.NullPointerException: Cannot invoke "com.sun.source.util.TreePath.getLeaf()" because "path" is null
        at com.palantir.suppressibleerrorprone.VisitorStateModifications.lambda$interceptDescription$0(VisitorStateModifications.java:44)
```

This is because we never find a top level type we can suppress, as we didn't think of anything but class as suppressible.

## After this PR
==COMMIT_MSG==
Don't fail to suppress on non-class type definitions (interfaces, enums, records etc)
==COMMIT_MSG==

I've also significantly improved the error message when this does go wrong:

```
public interface InterfaceType {
       ^
     Please report this at https://github.com/google/error-prone/issues/new and include the following:
  
     error-prone version: 2.31.0
     BugPattern: SafeLoggingPropagation
     Stack Trace:
     java.lang.RuntimeException: Can't find any source element on the TreePath to the error to place a @SuppressWarnings on. This is a bug an issues with suppressible-error-prone. The path to the error is:
  
  ===========================
  INTERFACE
  ===========================
  
  public interface InterfaceType {
      // ...
  }
  
  ===========================
  COMPILATION_UNIT
  ===========================
  package app;
  
  // ...
  
  public interface InterfaceType {
      // ...
  }
        at com.palantir.suppressibleerrorprone.VisitorStateModifications.lambda$interceptDescription$3(VisitorStateModifications.java:57)

```

### CODE REVIEW

File: suppressible-error-prone/src/main/java/com/palantir/suppressibleerrorprone/VisitorStateModifications.java [hunk 1]
```diff
@@ -44,10 +47,18 @@ public static Description interceptDescription(VisitorState visitorState, Descri
...
                 .dropWhile(path -> !suppressibleKind(path.getLeaf().getKind()))
                 .findFirst()
-                .orElseThrow(() -> new RuntimeException("Can't find anything we can suppress"))
+                .orElseThrow(() -> {
+                    return new RuntimeException("Can't find any source element on the TreePath to the error to "
+                            + "place a @SuppressWarnings on. This is a bug an with suppressible-error-prone. "
+                            + "The path to the error is:\n\n"
+                            + StreamSupport.stream(pathToActualError.spliterator(), false)
+                                    .map(tree -> tree.getKind().name() + "\n===========================\n" + tree)
+                                    .collect(Collectors.joining("\n\n")));
```
Comment (felixdesouza) [inline suppressible-error-prone/src/main/java/com/palantir/suppressibleerrorprone/VisitorStateModifications.java]: text block? 🥺

File: suppressible-error-prone/src/main/java/com/palantir/suppressibleerrorprone/VisitorStateModifications.java [hunk 2]
```diff
@@ -44,10 +47,18 @@ public static Description interceptDescription(VisitorState visitorState, Descri
...
 
-        Tree firstSuppressibleParent = Stream.iterate(pathToActualError, TreePath::getParentPath)
+        Tree firstSuppressibleParent = Stream.iterate(
+                        pathToActualError, treePath -> treePath.getParentPath() != null, TreePath::getParentPath)
                 .dropWhile(path -> !suppressibleKind(path.getLeaf().getKind()))
                 .findFirst()
-                .orElseThrow(() -> new RuntimeException("Can't find anything we can suppress"))
+                .orElseThrow(() -> {
+                    return new RuntimeException("Can't find any source element on the TreePath to the error to "
+                            + "place a @SuppressWarnings on. This is a bug an with suppressible-error-prone. "
```
Comment (felixdesouza) [inline suppressible-error-prone/src/main/java/com/palantir/suppressibleerrorprone/VisitorStateModifications.java]:
```suggestion
                            + "place a @SuppressWarnings on. This is a bug with suppressible-error-prone. "
```

File: suppressible-error-prone/src/main/java/com/palantir/suppressibleerrorprone/VisitorStateModifications.java (hunk 1 above)
Comment (CRogers) [reply suppressible-error-prone/src/main/java/com/palantir/suppressibleerrorprone/VisitorStateModifications.java]: I thought maybe but I don't think it's worth the bad formatting that palantir-java-format does without having actual newlines in the middle of the text.

Comment (felixdesouza) [reply suppressible-error-prone/src/main/java/com/palantir/suppressibleerrorprone/VisitorStateModifications.java]: i.e.

```
new RuntimeException("""
  Can't find any source element on the TreePath to the error to place a @SuppressWarnings on. This is a bug with suppressible-error-prone.

  The path to the error is:

  %s

  ===================

  %s
  """.formatted(StreamSupport.stream(...), tree);

potentially translated wrong but I think it works

Comment (CRogers) [reply suppressible-error-prone/src/main/java/com/palantir/suppressibleerrorprone/VisitorStateModifications.java]: We should probably fix PJF for this case:

```java
"""
Can't find any source element on the TreePath to the error to \
place a @SuppressWarnings on. This is a bug an with suppressible-error-prone.
The path to the error is:


"""
        + StreamSupport.stream(pathToActualError.spliterator(), false)
                .map(tree ->
                        tree.getKind().name() + "\n===========================\n" + tree)
                .collect(Collectors.joining("\n\n")));
```

should be

```java
"""
Can't find any source element on the TreePath to the error to \
place a @SuppressWarnings on. This is a bug an with suppressible-error-prone.
The path to the error is:


""" + StreamSupport.stream(pathToActualError.spliterator(), false)
        .map(tree ->
                tree.getKind().name() + "\n===========================\n" + tree)
        .collect(Collectors.joining("\n\n")));
```

imo

Comment (CRogers) [reply suppressible-error-prone/src/main/java/com/palantir/suppressibleerrorprone/VisitorStateModifications.java]: It also newlines + double indents `.formatted`.

Comment (felixdesouza) [reply suppressible-error-prone/src/main/java/com/palantir/suppressibleerrorprone/VisitorStateModifications.java]: true, but I would just do `"""...""".formatted(StreamSupport.stream())` and/or extract that to a variable. It is quite a gnarly expression to tack on

Comment (felixdesouza) [reply suppressible-error-prone/src/main/java/com/palantir/suppressibleerrorprone/VisitorStateModifications.java]: > It also newlines + double indents `.formatted`.

oof

File: suppressible-error-prone/src/main/java/com/palantir/suppressibleerrorprone/VisitorStateModifications.java (hunk 2 above)
Comment (CRogers) [reply suppressible-error-prone/src/main/java/com/palantir/suppressibleerrorprone/VisitorStateModifications.java]: fixed

File: suppressible-error-prone/src/main/java/com/palantir/suppressibleerrorprone/VisitorStateModifications.java (hunk 1 above)
Comment (CRogers) [reply suppressible-error-prone/src/main/java/com/palantir/suppressibleerrorprone/VisitorStateModifications.java]: Eh, I've done the text block version so if we fix PJF it should get autoformatted to something nice.

**Review by felixdesouza — APPROVED** [review]

**Pull request merged**